[Object,ELF] Implement PN_XNUM extension for program headers

aokblast opened the pull request:
In ELF file, there is a possible extended header for those phnum, shnum,
and shstrndx larger than the maximum of 16 bits. This extended header
use section 0 to record these fields in 32 bits. 

We implment this feature so that programs rely on ELFFile::program_headers() can get the
correct number of segments. Also, the consumers don't have to check the
section 0 themselve, insteead, they can use the getPhNum() as an
alternative.

jh7370 reviewed (changes requested):
The build is failing. Please fix it.

I've not looked at the change or the classes in depth, but the code already does try to handle large section counts, for example (see e.g. line 993 in ELF.h that you're trying to modify). What exactly are you trying to achieve that this code doesn't already do?

aokblast commented:
> Why do you want to support so many program headers? Just because the spec allows us doesn't mean we need to support it, if there's no use case for it.

It is related to the coredump design in FreeBSD. See: https://github.com/llvm/llvm-project/issues/132216.

MaskRay commented:
Please remove the lld/ELF change. For linker output we don't intend to support the PN_XNUM program header feature. We also don't accept features without a test.

Link: https://groups.google.com/g/generic-abi/c/-J3lNY8ZKkU ("PN_XNUM extension for program headers")

jh7370 reviewed (commented):
Thanks for the explanation. I agree that there's value in having the program header count extension in the parsing and dumping code.

In addition to removing the lld changes, I think this change warrants being split into at leasr two PRs, one (or more) which does the refactoring (if needed) such that both extended program headers and section headers can be handled all together and another that actually implements the program header functionality. You will also probably want a separate change specifically for llvm-objcopy, so that it can write objects containing many program headers.

aokblast commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -278,9 +278,17 @@ class ELFFile {
   ELFFile(StringRef Object);
 
 public:
+  const Elf_Word getPhNum() const { return e_phnum; }
```
Fixed. Thanks!

aokblast commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -889,15 +896,42 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+  // sections() parse the total number of sections by considering the
+  // extended headers
+  //
+  if (Result.getHeader().HasHeaderExtension()) {
```
The ELFFile::sections() and object::getSection() return an Expected. That means if you don't consume up the error, it will trigger panics by RAII. However, there is no way for us to pass error back from the constructor. The only way is through throw/catch. However, I don't  found any practice in llvm-prloject to use throw/catch so I think it is a bad idea for me to do so

aokblast commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -889,15 +896,42 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+  ELFFile Result(Object);
+
+  //
+  // sections() parse the total number of sections by considering the
```
Fixed. Thanks!

aokblast commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -278,9 +278,17 @@ class ELFFile {
   SmallString<0> FakeSectionStrings;
 
+  // Handle extended header in section 0
+  Elf_Word e_phnum;
+  Elf_Word e_shnum;
+  Elf_Word e_shstrndx;
```
Fixed. Thanks!

jh7370 commented on llvm/test/Object/invalid.test:
```diff
@@ -556,7 +556,7 @@ Sections:
 # RUN: not llvm-readobj -h %t25 2>&1 | FileCheck -DFILE=%t25 --check-prefix=INVALID-SEC-NUM1 %s
 
-# INVALID-SEC-NUM1: error: '[[FILE]]': unable to continue dumping, the file is corrupt: invalid section header table offset (e_shoff = 0x58) or invalid number of sections specified in the first section header's sh_size field (0x3ffffffffffffff)
+# INVALID-SEC-NUM1: error: '[[FILE]]': invalid section header table offset (e_shoff = 0x58) or invalid number of sections specified in the first section header's sh_size field (0x3ffffffffffffff)
```
Can we lazily populate the information?

One of the issues we've had in the past with llvm-readobj and other dumping tools is it would error out early because it couldn't read some part of the file, long before it actually needed to display the relevant file information, which meant in turn it became hard to impossible to diagnose the problem when a file was malformed in some way. In practice, when we dump the file headers, whether the entire section header table is readable is irrelevant, for example. We should only need to read the 0th section header to do certain things. I hope that makes sense.

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -889,7 +887,31 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+    return;
+
+  // An ELF binary may report `hasExtendedHeader` as true but not actually
+  // include an extended header. For example, a core dump can contain 65,535
```
The term "extended header" isn't used in the ELF gABI, so I don't think we should use it here. The header hasn't really been extended, in my opinion, so I think the term used in the Oracle docs (and possibly elsewhere) is misleading.

jh7370 reviewed (commented):
Please avoid force pushing to branches under active review, per LLVM policy, as it's unnecessary and means it's no longer possible to review changes to the PR since the previous review. You can use a merge commit, if you want to rebase, or fixup commits to make additional changes in the PR. All commits will be squashed into a single commit when landing the PR, because LLVM uses the "Sqaush & Merge" approach in GitHub. See https://llvm.org/docs/GitHub.html for more details.

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -889,7 +887,31 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+  if (!Header.hasPhdrNumExtension())
+    return;
+
+  // An ELF binary may report `hasExtendedHeader` as true but not actually
```
No, the getter should be the thing that does the initialization (if needed).

So something like:
```
mutable std::optional<uint32_t> RealPhNum;
mutable std::optional<uint32_t> RealShNum;
mutable std::optional<uint32_t> RealShStrndx;

Expected<uint32_t> getPhNum() const { // Probably should be defined in a .cpp at this point.
  if (!RealPhNum)
    if (Error E = readShdrZero())
      return std::move(E);
  return *RealPhNum;
}
// Same for ShNum and ShStrndx.

Error readShdrZero() {
  const Elf_Ehdr &Header = getHeader();
  if (Header.e_phnum == PN_XNUM || Header.e_shnum == 0 || Header.e_shstrndx == SHN_XINDEX)
  auto SecOrErr = getSection(0); // NB: maybe should use Expected<...> here instead, to fit LLVM coding standards.
  if (!SecOrErr)
    return SecOrErr.takeError();

  RealPhNum = Header.e_phnum == PN_XNUM ? (*SecOrErr)->sh_info : Header.e_phnum;
  RealShNum = Header.e_shnum == 0 ? (*SecOrErrr)->sh_size : Header.e_shnum;
  RealShStrndx = Header.e_shstrndx == SHN_XINDEX ? (*SecOrErr)->sh_link : Header.e_shstrndx;
}
```
Miscellaneous notes about the above:
- The above is not likely to be 100% correct, as I haven't tested it, but should help clarify what I'm expecting.
- The [Elf gABI](https://gabi.xinuos.com/) says `e_shnum == 0` indicates to use the 0th section header sh_size field, not SHN_UNDEF. While the two technically are the same value, it's more correct to compare against the literal 0 and not SHN_UNDEF.
- I've used PN_XNUM, not 0xffff, since we should use the constants when we have them and the gABI does (NB: the draft gABI hasn't been updated with the PN_XNUM extension yet, but it has been agreed on the mailing list).
- By doing this, the error isn't reported until the first (and subsequent) attempts to read the the real values, so if the values aren't needed, no error is reported, and the error is only reported once it is actually needed. If multiple reports of the error are undesirable, the caller can handle it (see e.g. llvm-readelf's `reportUniqueWarning`).
- If doing this, it's essential we don't reference `Real...` directly outside the example code I've posted, even if it's accessible. You could go as far as wrapping the fields and initialisation in another class, to enforce this, but I don't think it's worth doing that, personally.

aokblast replied:
Ok. I get what you mean.

A question is that should we consume up the error or report it back when the caller is inside the class?

Since some member functions like program_headers do not expect that it fails on reading e_phnum. That means we need to change many testcase if we want to report it back.

But you said that we don't want to see any difference in a user's viewpointl.

MaskRay replied:
Even if we don't want to introduce a test with 0xffff program headers, we can create a test, adapted from `test/llvm-objcopy/ELF/invalid-e_phoff.test`, that shows the llvm-readelf behavior when e_phnum is 0xffff, but the section header table doesn't contain that many of sections. The diagnostic is probably different.

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -887,30 +923,49 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+    if (!SecsOrErr) {
+      RealShNum = std::nullopt;
+      return SecsOrErr.takeError();
+    }
```
1) `sections()` should be modified to not do the dance with e_shnum/reading the index 0 header. The responsibility for this should be entirely in this function.

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -891,6 +925,49 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+    if ((*SecsOrErr).size() == 0) {
+      if (Header.e_phnum == ELF::PN_XNUM ||
+          Header.e_shstrndx == ELF::SHN_XINDEX) {
+        return createError("Unable to find Section 0");
```
I'm pretty sure it's a genuine failure, i.e. you've done something in this PR that is incorrect. For context, the failing test case is:

```
## Check the case when e_shstrndx == SHN_XINDEX, but null section's sh_link contains
## the index of a section header string table that is larger than the number of the sections.

# RUN: yaml2obj --docnum=30 %s -o %t30
# RUN: not llvm-objcopy %t30 2>&1 | FileCheck %s -DFILE=%t30 --check-prefix=INVALID-SHSTRTAB-INDEX

# INVALID-SHSTRTAB-INDEX: error: section header string table index 255 does not exist

--- !ELF
FileHeader:
  Class:     ELFCLASS64
  Data:      ELFDATA2LSB
  Type:      ET_REL
  Machine:   EM_X86_64
## SHN_XINDEX == 0xffff.
  EShStrNdx: 0xffff
Sections:
  - Type: SHT_NULL
    Link: 0xff
```

Section header 0 clearly should be readable in this case, but the test is failing with that unreadable section error you've introduced. The failure would come later on, when attempting to use the section header string table index (see https://github.com/llvm/llvm-project/blob/ec26f219acce77fb9b3d52abd31b0e639e788514/llvm/include/llvm/Object/ELF.h#L794). (Aside: `getSectionStringTable` and similar functions should be modified in this PR to reference `Real*` and call this method if the values are unitialised and not reference e_shnum directly).

aokblast replied:
I just realize that sh_link is 64bit field. Even gnu-readelf reports 32bit.
That is why I think it is a new error when I tried to compare the result with gnu version.

aokblast replied:
Fix comments.

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -887,30 +923,49 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+
+    // Pretend we have section 0 or sections() would call getShNum and thus
+    // become an infinite recursion
+    RealShNum = 0;
```
1) This code should be guarded by an `e_shnum == 0` check. If `e_shnum != 0`, we don't need to do funky things to read the index 0 header, like pre-filling `RealShNum` with a fake value, and can instead just set `RealShNum` to the right value up-front.

2) Perhaps this should be `1`, not `0`? That would allow `sections()` to read the first section header, if it is within the bounds of the file, or generate an appropriate error otherwise.

3) `sections()` should not be referencing `e_shnum` directly. It should use `RealShNum`, which will either contain the real value taken directly from e_shnum (because it's non-zero), a previously calculated value, derived from an earlier call to this function, or the fake value we set prior to calling `sections()` for the first time (i.e. 1 - see point 2)).

aokblast replied:
1. For your first question, we still need to read shdr 0 when phnum is 65535.

2. If we actually have sh_num == 0, and both phnum and shstrndx are valid. It should not be considered as an Error. Right? Since the search of shdr 0 is caused by shnum (not phnum or shstrndx) but we actually have 0 sections. For this case, we should not regard it as an error.

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -891,6 +924,39 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+      Header.e_shoff != 0) {
+
+    // Pretend we have section 0 or sections() would call getShNum and thus
+    // become an infinite recursion
```
```suggestion
    // become an infinite recursion.
```
Nit: comments should be full sentences, with trailing full stops.

jh7370 reviewed (commented):
Thanks for the updates. This is looking mostly good. I've got a few remaining small comments.

We'll need to audit the existing code within LLVM, to ensure people aren't doing the section 0 reading themselves still, but those can probably be left until future PRs, if the tests are all passing.

aokblast commented:
Sorry for the late reply and thanks for continously reviewing my patches!
Some style nit are really hard for me to get rid of as a developer of other projects.

> Thanks for the updates. This is looking mostly good. I've got a few remaining small comments.
> 
> We'll need to audit the existing code within LLVM, to ensure people aren't doing the section 0 reading themselves still, but those can probably be left until future PRs, if the tests are all passing.

Sure. It is now in my TODO list. Thanks for your advise! I would check that after the readobj patch is done.

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -891,6 +927,38 @@ Expected<uint64_t> ELFFile<ELFT>::getDynSymtabSize() const {
+    if (Header.e_shnum == 0)
+      RealShNum = 1;
+    else
+      RealShNum = Header.e_shnum;
```
This can be simplified to simply assigning 1 to `RealShNum` unconditionally. The code that assigns `RealShNum` below will fix it to the right value afterwards, regardless of whether `Header.e_shnum == 0` or otherwise.

jh7370 reviewed (approved):
LGTM, on the proviso that we add testing for the new program header code in a follow-up PR, e.g. when updating llvm-readobj or another tool.

aokblast committed:
[Object,ELF] Implement PN_XNUM extension for program headers

In ELF file, there is a possible extended header for those phnum, shnum,
and shstrndx larger than the maximum of 16 bits. This extended header
use section 0 to record these fields in 32 bits.

We implment this feature so that programs rely on ELFFile::program_headers() can get the
correct number of segments. Also, the consumers don't have to check the
section 0 themselve, insteead, they can use the getPhNum() as an
alternative.
